src/net/download.rs

```diff
@@ -0,0 +1,1859 @@
+            blocks_to_try: HashMap::new(),
+            microblocks_to_try: HashMap::new(),
+
+            parsed_urls: HashMap::new(),
+            dns_lookups: HashMap::new(),
+            dns_timeout: dns_timeout,
+
+            getblock_requests: HashMap::new(),
+            getmicroblocks_requests: HashMap::new(),
+            blocks: HashMap::new(),
+            microblocks: HashMap::new(),
+
+            dead_peers: vec![],
+            broken_peers: vec![],
+            broken_neighbors: vec![],
+        }
+    }
+
+    pub fn reset(&mut self) -> () {
+        self.state = BlockDownloaderState::DNSLookupBegin;
+
+        self.dns_lookups.clear();
+        self.parsed_urls.clear();
+
+        self.getblock_requests.clear();
+        self.getmicroblocks_requests.clear();
+        self.blocks_to_try.clear();
+        self.microblocks_to_try.clear();
+        self.blocks.clear();
+        self.microblocks.clear();
+
+        self.dead_peers.clear();
+        self.broken_peers.clear();
+        self.broken_neighbors.clear();
+
+        // perserve sortition height
+        // preserve download accounting
+    }
+
+    pub fn restart_scan(&mut self) -> () {
+        // prepare to restart a full-chain scan for block downloads
+        self.block_sortition_height = 0;
+        self.microblock_sortition_height = 0;
+        self.next_block_sortition_height = 0;
+        self.next_microblock_sortition_height = 0;
+        self.empty_block_download_passes = 0;
+        self.empty_microblock_download_passes = 0;
+    }
+
+    pub fn dns_lookups_begin(&mut self, dns_client: &mut DNSClient, mut urls: Vec<UrlString>) -> Result<(), net_error> {
+        assert_eq!(self.state, BlockDownloaderState::DNSLookupBegin);
+
+        self.dns_lookups.clear();
+        for url_str in urls.drain(..) {
+            let url = url_str.parse_to_block_url()?;        // NOTE: should always succeed, since a UrlString shouldn't decode unless it's a valid URL
+            let port = match url.port_or_known_default() {
+                Some(p) => p,
+                None => {
+                    warn!("Unsupported URL {:?}: unknown port", &url);
+                    continue;
+                }
+            };
+            match url.host() {
+                Some(url::Host::Domain(domain)) => {
+                    dns_client.queue_lookup(domain.clone(), port, get_epoch_time_ms() + self.dns_timeout)?;
+                    self.dns_lookups.insert(url_str.clone(), None);
+                    self.parsed_urls.insert(url_str, DNSRequest::new(domain.to_string(), port, 0));
+                },
+                Some(url::Host::Ipv4(addr)) => {
+                    self.dns_lookups.insert(url_str, Some(vec![SocketAddr::new(IpAddr::V4(addr), port)]));
+                }
+                Some(url::Host::Ipv6(addr)) => {
+                    self.dns_lookups.insert(url_str, Some(vec![SocketAddr::new(IpAddr::V6(addr), port)]));
+                },
+                None => {
```

Comment:
How does a node behave if it receives a URL such as `localhost`, `127.0.0.1`, etc?

Reply:
It works.  The tests all do this.